arcv: Add APEX support.

luismgsilva opened the pull request:
Thanks for taking the time to contribute to GCC! Please be advised that if you are
viewing this on `github.com`, that the mirror there is unofficial and unmonitored.
The GCC community does not use `github.com` for their contributions. Instead, we use
a mailing list (`gcc-patches@gcc.gnu.org`) for code submissions, code reviews, and
bug reports. Please send patches there instead.

Commit:
arcv: apex: Add `#pragma intrinsic` handler.

Implements support for a new pragma of the form:

    #pragma intrinsic(fn_name, "insn_name", opcode, "format"...)

This pragma registers a user-defined intrinsic to be treated
as built-in APEX instructions during compilation.  Parsing
its name, instruction mnemonic, opcode and a set of optional
format specifiers ("XD", "XS", "XI", "XC").

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: set APEX operand flags based on function prototype.

Adds `arcv_apex_set_insn_operand_flags`, which examines the return type
and arguments of a built-in function to determine the appropriate APEX
operand flags.  A non-void return type sets APEX_DEST, and up to two
arguments set APEX_SRC0 and APEX_SRC1.

Emits an error if the function has more than two arguments.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Determine APEX instruction format when unspecified.

Adds `arcv_apex_resolve_insn_format`, which computes a concrete APEX
format (XD, XS, XI or XC) when no explicit format is provided. The
selection is based on the operand configuration and opcode value.

Also introduces `APEX_OPCODE_FIELD_MAX` and operand signature masks
to guide format resolution based on accepted operand patterns.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add validation for insn. format and operand rules.

Introduces `arcv_apex_validate_insn_format`, which checks whether a
specified APEX instruction format is valid given the opcode and
operand layout. The validation includes:

- Ensuring the opcode falls within the allowed range for the format.
- Verifying the number of scalar operands is appropriate.
- Enforcing format-specific constraints such as return value presence.

The supported format rules are encoded in a `format_rule` table,
covering XD, XS, XI and XC.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Print ".extInstruction" assembly section.

Adds `arcv_apex_print_insn_section`, which prints
`.extInstruction` assembly sections for APEX instructions.
This function determines the appropriate formats and operand
annotations to output based on the insn_format bitmask.

The output helps the assembler and disassembler recognize supported
instruction formats, including operand roles (void, no_src0, no_src1)
and format classes (XD, XS, XI, XC).

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add machine description patterns for APEX instructions.

Defines new `define_insn` patterns to support ARC-V APEX instructions
across all supported formats (`XI`, `XS`, `XC`, `XD`).  These patterns
emit custom assembly using the `arcv_apex_get_insn_name` hook to
dynamically resolve the instruction mnemonic from the subcode operand.

Covers the following instruction forms:

- `insn`: used by "XD" (no operands)
- `insn src0`: used by "XI","XD" (immediate or register)
- `insn src0, src1`: used by "XS","XD" (register or mixed with immediate)
- `insn dest`: used by "XD"
- `insn dest, src0`: used by "XI","XD"
- `insn dest, src0, imm/src1`: used by "XS","XC","XD"

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add constraints and format support check.

Introduce new ARC-V APEX constraints to validate operand values and
instruction format support for user-defined builtins:

  - Add constraints "B8" for 8-bit signed immediates (-128 to 127).
  - Add format-specific constraints "xAVpXD", "xAVpXS", "xAVpXI" and
    "xAVpXC" to check whether a given APEX instruction (by subcode)
    supports the respective instruction format.

Also implement "arcv_apex_format_supports_p ()" in "riscv-builtins.cc",
which performs a bitmask check against the stored instruction format
flags for a given APEX builtin. Used during instruction selection to
ensure only applicable patterns match.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add lookup function declaration helper.

Introduce `arcv_apex_lookup_function` to return the function declaration
corresponding to a string name provided via the #pragma intrinsic
directive.  It performs a scoped lookup, verifies that the name refers
to a function, and issues an error if not declared.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add subcode operand to APEX RTX instructions.

This patch extends "riscv_expand_builtin_direct" to optionally accept
and attach a subcode operand for APEX builtins. A new "has_subcode_p"
flag determines whether a constant RTL operand for the subcode should
be added to the operand list.

The subcode is used to extract the instruction name from
"arcv_apex_builtins" once a pattern is selected in the machine description
(see "arcv_apex_get_insn_name (rtx op)").

Also updates "riscv_check_builtin_call" and
"riscv_resolve_overloaded_builtin" to handle the new
"RISCV_BUILTIN_APEX" class.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Map APEX operand layouts to insn codes.

Adds `arcv_apex_get_icode`, a helper that selects the appropriate
`insn_code` based on the operand layout encoded in `insn_format`.
It distinguishes among patterns like `void`, `src0`, `dest,src0`
and `dest,src0,src1` and maps them to the corresponding
`CODE_FOR_riscv_arcv_apex_*` enum used in RTL generation.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add APEX intrinsic built-in registration mechanism.

Introduce infrastructure to register user-defined APEX intrinsics as
RISC-V built-in functions during pragma handling.  This includes:

- Implementing arcv_apex_init_builtin(), which validates and resolves
  its instruction format, determines its internal instruction code
  (icode), modifies an existing function declaration as a built-in
  function, and updates the function declaration with custom function
  code.
- Hooking up arcv_apex_init_builtin() in arcv_apex_pragma_intrinsic().
- Extending riscv_builtin_class with RISCV_BUILTIN_APEX and updating
  RISCV_BUILTIN_SHIFT to 2.

This allows the compiler to recognize and process APEX intrinsics
defined by users via pragma directives as first-class built-in functions.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add volatile variants for instructions with side effects.

Introduce "unspec_volatile" to handle APEX instructions that have
side effects and must not be optimized away by the compiler.
When an instruction is defined with the "side_effect" attribute via
preventing optimizations such as common subexpression elimination,
loop-invariant motion, or dead code elimination.

Instructions with no destination register are treated as volatile.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Register pragma handler after macro expansion.

Use "c_register_pragma_with_expansion()" instead of "c_register_pragma()" to
ensure that "arcv_apex_pragma_intrinsic" is invoked with arguments that
have been macro-expanded.

This allows users to define APEX intrinsic names or parameters using
macros in pragma directives, ensuring the handler processes the
expanded values  rather than raw macro identifiers.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Use recog-based selection of variants via mode iterators.

GCC selects the "define_expand" based on the operand layout of the
function declaration.
Since the declaration may use different orders or types of operands, we
take full advantage of GCC’s internal instruction recognizer to match
RTL expressions with the correct pattern in the machine description.

During the "define_expand", the compiler generates a generic RTL using
an "UNSPEC" code to uniquely identify a family of custom instructions
along with their operands.
At this stage, it's not yet a real instruction — it's an abstract
placeholder tagged with a unique "UNSPEC" code.

Later, during instruction selection, GCC calls "recog", which matches
this RTL expression to the appropriate "define_insn" based on:

- The "UNSPEC" code (acts as a unique identifier);
- Operand modes (e.g., "SI", `DI`, etc.);
- Operand predicates and constraints.

If all conditions match, GCC selects the correct "define_insn", and code
generation proceeds accordingly.

To cover all valid combinations of operand types, the "define_insn" uses
mode iterators (like "DM", "S0M", "S1M"), allowing it to generate all
required mode variants from a single pattern template.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Validate immediate arguments during expansion.

Add validation for immediate arguments in APEX intrinsic calls
that do not support the APEX_XD format. This check is performed
during riscv_expand_builtin_direct, where the actual values of
the current RTL expression are available.

The validation ensures that the argument is a constant integer
and falls within the correct range:

- For APEX_XI and APEX_XC formats: signed 12-bit [-2048, 2047]
- For APEX_XS format: signed 8-bit [-128, 127]

Report an error if the value does not meet the previous checks.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add "i" suffix to resolved immediate-format instructions.

When an APEX instruction does not have a user-defined format specified
via pragma, its format is resolved based on operand flags and opcode.  In
these cases, this patch appends an "i" suffix to the instruction name to
distinguish it from the XD format.  While XD instructions and
immediate-based formats (XI, XS, XC) may have the same number of
operands, XD uses only registers, whereas the others include at least
one immediate operand.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Avoid registering instructions with duplicate opcode and insn format.

Add a check to detect when a new APEX intrinsic specifies an opcode
value that overlaps with an already registered intrinsic, considering
only the lower 4 bits of the insn format (XD, XS, XI, XC).  Emit an
error in such cases to prevent opcode conflicts at runtime.

Signed-off-by: Luis Silva <luiss@synopsys.com>

MichielDerhaeg commented on gcc/config/riscv/riscv-c.cc:
```diff
@@ -100,6 +101,235 @@ riscv_pragma_intrinsic_flags_restore (struct pragma_intrinsic_flags *flags)
+      insn_name_raw = IDENTIFIER_POINTER (x);
+      break;
+    default:
+      error ("pragma intrinsic: APEX attribute 'name' is missing");
```
Missing return?

luismgsilva replied:
In some cases, we want the compiler to keep executing so it can report all errors related to the pragma syntax at once.

MichielDerhaeg commented on gcc/config/riscv/riscv-c.cc:
```diff
@@ -100,6 +101,235 @@ riscv_pragma_intrinsic_flags_restore (struct pragma_intrinsic_flags *flags)
+  const char *fn_name = IDENTIFIER_POINTER (x);
+
+  /* Consume the comma.  */
+  pragma_lex (&x);
```
Check for comma, could be anything.

luismgsilva replied:
The intention was for the compiler to report all missing attributes without considering the presence of a comma. That’s how CCAC behaves.

For example:
 ```c
int foo(int);

#pragma intrinsic(foo) 
```
It complains with:
```bash
arcv-apex-err5.c:6:9: error: pragma intrinsic: APEX attribute 'name' is missing
    6 | #pragma intrinsic(foo)
      |         ^~~~~~~~~
arcv-apex-err5.c:6:9: error: pragma intrinsic: APEX attribute 'opcode' is missing
```

When we add extra validations for the comma, we end up with output like:
```bash
arcv-apex-err5.c:6:9: error: pragma intrinsic: expected ',' after APEX attribute name 'foo'
    6 | #pragma intrinsic(foo)
      |         ^~~~~~~~~
arcv-apex-err5.c:6:9: error: pragma intrinsic: APEX attribute 'name' is missing
arcv-apex-err5.c:6:9: error: pragma intrinsic: expected ',' after APEX attribute name ''
arcv-apex-err5.c:6:9: error: pragma intrinsic: APEX attribute 'opcode' is missing
```

If we’re okay with having this difference in behavior between the two compilers, I have no objection.

MichielDerhaeg replied:
It does look cleaner indeed if we omit the comma error.
But if we don't check for it, it will parse it wrong if you forget about the comma.

Can we check for it, but delay the error, and omit it when it makes sense?

MichielDerhaeg replied:
And if you really don't want to check it, at least document why.

luismgsilva replied:
Since omitting the comma always leads to a missing attribute error anyway, reporting a separate comma error would be redundant. We could technically validate the comma during parsing and only report it if no other attribute-related errors are found - though in practice, that situation doesn’t occur.
For clarity, I added a comment in the source code to document the reasoning.

MichielDerhaeg commented on gcc/config/riscv/riscv-c.cc:
```diff
@@ -100,6 +101,235 @@ riscv_pragma_intrinsic_flags_restore (struct pragma_intrinsic_flags *flags)
+  }
+
+  /* Reject instruction names that do not follow APEX identifier rules.  */
+  if (insn_name_raw[0] != '\0'
```
And what if it is the empty string?

luismgsilva replied:
Good catch. Fixed.

MichielDerhaeg commented on gcc/config/riscv/riscv-c.cc:
```diff
@@ -100,6 +101,235 @@ riscv_pragma_intrinsic_flags_restore (struct pragma_intrinsic_flags *flags)
+    *p = TOLOWER (*p);
+
+  /* Consume the comma.  */
+  pragma_lex (&x);
```
Idem.

MichielDerhaeg commented on gcc/config/riscv/arcv-apex.md:
```diff
@@ -0,0 +1,437 @@
+
+;; Used by "XD" insn. format: `insn`
+(define_insn "riscv_arcv_apex_void_ftype_v"
+  [(unspec_volatile:SI [(match_operand:SI 0 "const_int_operand" "xAVpXD")]
```
How is it determined whether this needs to be volatile? I know it is specified in the syntax, but how is the right insn picked?

luismgsilva replied:
If an instruction is defined with a void return type and no arguments, then that’s the only valid form it can take. See:
https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/blob/861b8c603897e88c16c563e55764cf82d18ff124/gcc/config/riscv/riscv-builtins.cc#L905

Any instruction without a return value (and therefore no destination register) is marked as volatile.

MichielDerhaeg commented on gcc/config/riscv/riscv-builtins.cc:
```diff
@@ -369,6 +473,17 @@ riscv_expand_builtin_direct (enum insn_code icode, rtx target, tree exp,
+    /* Add the subcode as an additional input operand to the RTL expression.  */
+    create_input_operand (&ops[opno++], const_rtx, SImode);
+    /* Validate the immediate argument passed to the APEX intrinsic.  */
+    if (!arcv_apex_immediate_argument_valid_p (subcode, exp))
+      return const0_rtx;
```
What happens if it isn't valid?

luismgsilva replied:
The validations happen inside the `arcv_apex_immediate_argument_valid_p` function. If an error is triggered, the caller simply exits by returning const0_rtx.

MichielDerhaeg commented on gcc/config/riscv/riscv-builtins.cc:
```diff
@@ -420,16 +535,33 @@ riscv_expand_builtin (tree exp, rtx target, rtx subtarget ATTRIBUTE_UNUSED,
       case RISCV_BUILTIN_VECTOR:
 	return riscv_vector::expand_builtin (subcode, exp, target);
+      case RISCV_BUILTIN_APEX: {
+  const struct arcv_apex_builtin_description *d = &arcv_apex_builtins[subcode];
+
+  switch (d->builtin_type)
+    {
+    case RISCV_BUILTIN_DIRECT:
+      return riscv_expand_builtin_direct (d->icode, target, exp, true,
+					  subcode, true);
+
+    case RISCV_BUILTIN_DIRECT_NO_TARGET:
+      return riscv_expand_builtin_direct (d->icode, target, exp, false,
+					  subcode, true);
+    }
+    break;
```
Formatting looks a bit off here.

luismgsilva replied:
I kept it close to the existing pattern in `riscv_expand_builtin` to minimize stylistic discrepancies

MichielDerhaeg replied:
Everything inside the `RISCV_BUILTIN_APEX` needs to be indented a bit more.

luismgsilva replied:
Fixed. Thanks.

MichielDerhaeg commented on gcc/config/riscv/riscv-builtins.cc:
```diff
@@ -454,3 +586,421 @@ riscv_atomic_assign_expand_fenv (tree *hold, tree *clear, tree *update)
+    tree type_to_check = argtype;
+    if (TREE_CODE (argtype) == POINTER_TYPE)
+    {
+      /* Skip size check if return type is void.  */
+      if (return_type == void_type_node)
+	continue;
```
Why does the return type influence anything if the arg type is a pointer?

MichielDerhaeg replied:
I just read this:
> APEX function %qs must return the same type as the first parameter for the %qs format class

But in this case they are not the same?

luismgsilva replied:
> Why does the return type influence anything if the arg type is a pointer?

If an intrinsic has a void return type, it can accommodate a larger size.

luismgsilva replied:
> I just read this:
> 
> > APEX function %qs must return the same type as the first parameter for the %qs format class
> 
> But in this case they are not the same?

For "XC" instruction format the return type and first argument must be the same type. Because the destination register and first argument operand are the same register.

MichielDerhaeg commented on gcc/config/riscv/riscv-builtins.cc:
```diff
@@ -454,3 +586,421 @@ riscv_atomic_assign_expand_fenv (tree *hold, tree *clear, tree *update)
+  fndecl->function_decl.function_code
+	= (arcv_apex_builtin_index << RISCV_BUILTIN_SHIFT) + RISCV_BUILTIN_APEX;
+
+  arcv_apex_builtin_index++;
```
More like a size/amount than an index, no?

luismgsilva replied:
I understand your point regarding the naming. `arcv_apex_builtin_index` is used both as an index for inserting the next entry and as the current size for iteration bounds. If there’s a preference for clarity, I’m fine renaming it to `arcv_apex_builtin_count`. Otherwise, I kept index since its primary role is writing into the array.

MichielDerhaeg replied:
I was just nitpicking.

MichielDerhaeg commented on gcc/config/riscv/riscv-builtins.cc:
```diff
@@ -454,3 +586,421 @@ riscv_atomic_assign_expand_fenv (tree *hold, tree *clear, tree *update)
+const char*
+arcv_apex_get_insn_name (rtx op)
+{
+	unsigned int subcode = UINTVAL (op);
```
Formatting. Is now a tab, should be 2 spaces.

luismgsilva replied:
Fixed. Thanks

luismgsilva commented:
The most recent force push addressed the comments in the emitted instructions. For RISC-V, the standard assembly comment symbol is `#`, whereas for ARC it is `;`.

Commit:
arcv: apex: Add builtin description and registry table.

Introduce "arcv_apex_builtin_description" to describe ARC-V APEX
builtins, including instruction code, builtin name, instruction name,
instruction opcode, expansion type, and instruction format flags.

Add global "arcv_apex_builtins[]" array and index variable to register
builtins at compile time. Define a builtin limit of 384 entries, based
on the encoding capacity of XD (256), XS (64), XI (32), and XC (32)
instruction formats.

Signed-off-by: Luis Silva <luiss@synopsys.com>

Commit:
arcv: apex: Add ARC-V APEX Compiler testsuite.

This testsuite tests the compiler to generate the correct
APEX instruction formats and their respective variants.

Also validates error handling.

Signed-off-by: Luis Silva <luiss@synopsys.com>
